File: frontend/src/objects/user/response/ManagedUserListItem.ts
import type { DisplayName } from '@/objects/user/DisplayName'
import type { Username } from '@/objects/user/Username'
import type { EmailAddress } from '@/objects/auth/EmailAddress'

/** 管理员可见的用户列表项；包含邮箱和权限位，不能用于公开资料页。 */
export type ManagedUserListItem = {
  username: Username
  displayName: DisplayName
  email: EmailAddress
  siteManager: boolean
  problemManager: boolean
  contestManager: boolean
}

File: backend/src/main/scala/domains/user/objects/response/ManagedUserListItem.scala
package domains.user.objects.response

import domains.user.objects.*

import domains.auth.objects.EmailAddress
import domains.user.objects.{DisplayName, Username}
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}

/** 管理端用户列表项，合并用户资料、邮箱和权限状态。 */
final case class ManagedUserListItem(
  username: Username,
  displayName: DisplayName,
  email: EmailAddress,
  siteManager: Boolean,
  problemManager: Boolean,
  contestManager: Boolean
)

/** 提供管理端用户列表项 JSON 编解码。 */
object ManagedUserListItem:
  given Encoder[ManagedUserListItem] = deriveEncoder[ManagedUserListItem]
  given Decoder[ManagedUserListItem] = deriveDecoder[ManagedUserListItem]
